Source/data/common.hpp

```diff
@@ -0,0 +1,226 @@
+#pragma once
+
+#include <memory>
+
+#include "engine.h"
+#include "utils/stdcompat/algorithm.hpp"
+#include "utils/stdcompat/cstddef.hpp"
+#include "utils/stdcompat/string_view.hpp"
+
+namespace devilution {
+enum class DataFileError {
+	NotFound = 1,
+	ReadError
+};
+
+struct LoadDataResult {
+	std::unique_ptr<char[]> start;
+	size_t size;
+	DataFileError ec;
+
+	LoadDataResult() = delete;
+
+	LoadDataResult(std::unique_ptr<char[]> &&start, size_t size)
+	    : start(std::move(start))
+	    , size(size)
+	    , ec()
+	{
+	}
+
+	LoadDataResult(DataFileError ec)
+	    : start(nullptr)
+	    , size(0)
+	    , ec(ec)
+	{
+	}
+
+	[[nodiscard]] bool hasError()

```

Comment:
Let's use `has_value()` to mimic `std::expected` from C++23.
I think we should add an `std::expected` backport at this point since we have this pattern in quite a few places (I'll work on this in a separate PR).